lib/dat-tcp.rb

```diff
@@ -4,37 +4,40 @@
 
 require 'dat-tcp/version'
 require 'dat-tcp/logger'
+require 'dat-tcp/worker'
 
 module DatTCP
 
   class Server
 
-    attr_reader :worker_start_procs, :worker_shutdown_procs
-    attr_reader :worker_sleep_procs, :worker_wakeup_procs
+    DEFAULT_NUM_WORKERS = 2
+
     attr_reader :logger
     private :logger
 
-    def initialize(config = nil, &serve_proc)
-      config ||= {}
-      @backlog_size     = config[:backlog_size]     || 1024
-      @debug            = config[:debug]            || false
-      @min_workers      = config[:min_workers]      || 2
-      @max_workers      = config[:max_workers]      || 4
-      @shutdown_timeout = config[:shutdown_timeout] || 15
-      @signal_reader, @signal_writer = IO.pipe
-      @serve_proc = serve_proc || raise(ArgumentError, "no block given")
+    def initialize(worker_class, options = nil)
+      if !worker_class.kind_of?(Class) || !worker_class.include?(DatTCP::Worker)
+        raise ArgumentError, "worker class must include `#{DatTCP::Worker}`"
+      end
 
-      @worker_start_procs    = []
-      @worker_shutdown_procs = []
-      @worker_sleep_procs    = []
-      @worker_wakeup_procs   = []
+      options ||= {}
+      @backlog_size     = options[:backlog_size]     || 1024
+      @debug            = options[:debug]            || false
+      @shutdown_timeout = options[:shutdown_timeout] || 15
+
+      @signal_reader, @signal_writer = IO.pipe
 
       @logger = DatTCP::Logger.new(@debug)
 
-      @tcp_server       = nil
-      @work_loop_thread = nil
-      @worker_pool      = nil
-      @signal = Signal.new(:stop)
+      options[:num_workers] ||= DEFAULT_NUM_WORKERS
+      @worker_pool = DatWorkerPool.new(worker_class, {
+        :num_workers   => options[:num_workers],
```

Comment:
@jcredding wouldn't it be better to do `:num_workers   => options[:num_workers] || DEFAULT_NUM_WORKERS,` here and not mess with the passed in options hash on line 32?  Plus more consistent when option defaulting above.